### ISSUE TITLE
New maximum character limit for tweets. Suggestions welcome.

### ISSUE THREAD

**Issue by dentoir** [digitalmethodsinitiative/dmi-tcat#228]
There will be an increased maximum character limit for tweets next week (for entities, not the plain text), as discussed in the links below. It would be greate to have an in-place workaround before 19th of September, and a longterm solution.

http://www.theverge.com/2016/9/12/12891562/twitter-tweets-140-characters-expand-photos
https://blog.twitter.com/express-even-more-in-140-characters

Currenty the text column in the database is a `VARCHAR(255)`. This leaves space enough for more information, but the new recommendation is to no longer limit the storage size and we can expect at least a portion of all tweets to exceed that size (difficult to estimate how many).

URLs, media files and mentions are extracted from the JSON data as delivered by the Twitter API. This functionality will not break (though it is worthwhile to revisit the code and confirm that). Truncating the tweet text to 255 characters would break searching users and URLs _inside the tweet text_.

Any suggestions on how to fix this, taking into account efficiency, upgradability, etc.? Is it feasible/desirable to transform the tweet text column to TEXT, or to just increase the VARCHAR dimensions, for example?

**Comment by ErikBorra** [comment]
In addition: mentions at the beginning and URLs (possibly to images) at the end will not count towards the 'official' 140 char limit. Rather they will be denoted as 'hidden' in the JSON output. One way forward would thus be not to store the 'hidden' parts in the text column, as this (additional) info is already stored in the mention and URL tables.

**Comment by dentoir** [comment]
Hi all,

This document describes in detail what kind of changes are expected and what the default return behavior of the API shall be for the moment. In the appendix you can find several JSON files. 

https://dev.twitter.com/overview/api/upcoming-changes-to-tweets

I've been thinking about what would need to change and what could be affected. For the end-user of TCAT this would probably only affect their ability to search for (parts of) URLs inside the tweet text. There is also a potential impact on our fix in issue #74 (truncation when retweeting) which may become redundant.

In any case, I don't see how this necessitates major changes. @hoylen and @brendam, do you have any thoughts on this? I've made a new test branch, https://github.com/digitalmethodsinitiative/dmi-tcat/tree/extended

Currently this branch:

- stores any full tweet text up till 255 characters
- truncates the tweet text till the 140 displayed characters if the full text > 255 characters and sets the truncated boolean to true

**Comment by hoylen** [comment]
Knowing whether a tweet has been truncated or not is important. That way you know you are missing data, if it was truncated.

Changing the _varchar_ to a _text_ field is certainly the more "lossless" approach. But is a backward-incompatible change; would require lots of testing to make sure nothing breaks; and might have significant storage size and performance implications on what is already a large database. Maybe some testing on what that storage and performance impact is might convince you one way or the other.

I don't have any opinion about whether truncating or not is the best solution. Obviously "lossless" sounds better than "lossy", but has to be weighed up against cost (implementation effort and performance).

**Comment by dentoir** [comment]
I've made a branch which, simply put, stores as much in the 'text' field as possible, and when the internal storage limit is reached, truncates and sets the truncate flag to true.

https://github.com/digitalmethodsinitiative/dmi-tcat/tree/longertweets

I'm testing this on our servers this weekend. I don't expect any problems and will merge with master after the weekend.

In the meantime, I discovered that when a tweet contains many UTF8MB4 characters, it can _already_ extend beyond the 140 bytes. Apparantly a VARCHAR() element != an UTF8MB4 character. This was a previously unknown issue.

The following tweet will be truncated in TCAT: https://twitter.com/DJBOBBY30

For a bin tracking 'Paris', we got 0.006% reaching the internal storage limit already. Therefore, we decided to experiment with two possible alterations to the storage: either set tweet to TEXT or increase the VARCHAR() size.
